runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -59,16 +250,175 @@ extern "C" {
+
+    // Test the return type
+    switch (sigArray[lastSigIdx].type & J9_FFI_UPCALL_SIG_TYPE_MASK) {
+        case J9_FFI_UPCALL_SIG_TYPE_VOID:
+            metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcall0;
+            break;
+        case J9_FFI_UPCALL_SIG_TYPE_CHAR:
+        case J9_FFI_UPCALL_SIG_TYPE_SHORT:
+        case J9_FFI_UPCALL_SIG_TYPE_INT32:
+            metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcall1;
+            break;
+        case J9_FFI_UPCALL_SIG_TYPE_POINTER:
+        case J9_FFI_UPCALL_SIG_TYPE_INT64:
+            metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallJ;
+            break;
+        case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+            metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallF;
+            break;
+        case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+            metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallD;
+            break;
+        default:
+            Assert_VM_unreachable();
+    }
+
+    I_32 gprArgCount = 0;
+    I_32 fprArgCount = 0;
+    // We have 3 GPRs (R1 to R3) and 4 FPRs (F0, F2, F4, F6) available for parameter passing.
+    const I_32 maxGPRArgCount = 3;
+    const I_32 maxFPRArgCount = 4;
+
+    // Loop through the arguments and count how many instructions you'll need to copy data from registers to memory, and how many
+    // bytes of space you'll need to allocate in thunk and new stack frame to store the instructions + parameter info.
+    for (I_32 i = 0; i < lastSigIdx; i++) {
+        I_32 argSizeInBytes = sigArray[i].sizeInByte;
+        switch (sigArray[i].type) {
+            case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT64:
+                if (gprArgCount < maxGPRArgCount) {
+                    numInstructions++;
+                }
+                gprArgCount++;
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+                if (fprArgCount < maxFPRArgCount) {
+                    numInstructions++;
+                }
+                fprArgCount++;
+                break;
+            default:
+                Assert_VM_unreachable();
+        }
+        if (fprArgCount + gprArgCount >= 7)
+            break;
+
+    }
+
+    // The thunk will frame will be 256 bytes in size. Since parameters will be stored in the caller frame, offsetFromR4
+    // must account for the frame size of the thunk.
+    const I_32 defaultStackPointerBias = 2048;
+    const I_32 offsetToParamArea = 128;
+    const I_32 calleeFrameSize = 256;
+    const I_32 offsetFromR4 = defaultStackPointerBias + offsetToParamArea + calleeFrameSize;
```

Comment:
Can we use a better name here instead of offsetFromR4?

Reply:
This should be fixed now in [968f54e](https://github.com/eclipse-openj9/openj9/pull/19701/commits/968f54e09c9943b50bacbc4329189a3ef85ec21f). I renamed it to `argAreaOffsetFromSP`.